File: projectforge-business/src/main/java/org/projectforge/business/timesheet/TimesheetDO.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.timesheet;

import java.sql.Timestamp;
import java.util.Date;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import javax.persistence.Transient;

import org.apache.commons.lang.StringUtils;
import org.hibernate.search.annotations.Analyze;
import org.hibernate.search.annotations.DateBridge;
import org.hibernate.search.annotations.EncodingType;
import org.hibernate.search.annotations.Field;
import org.hibernate.search.annotations.Index;
import org.hibernate.search.annotations.Indexed;
import org.hibernate.search.annotations.IndexedEmbedded;
import org.hibernate.search.annotations.Resolution;
import org.hibernate.search.annotations.Store;
import org.projectforge.business.fibu.kost.Kost2ArtDO;
import org.projectforge.business.fibu.kost.Kost2DO;
import org.projectforge.business.task.TaskDO;
import org.projectforge.framework.persistence.entities.DefaultBaseDO;
import org.projectforge.framework.persistence.user.api.UserPrefParameter;
import org.projectforge.framework.persistence.user.entities.PFUserDO;
import org.projectforge.framework.time.DateHolder;
import org.projectforge.framework.time.DatePrecision;
import org.projectforge.framework.time.DateTimeFormatter;
import org.projectforge.framework.time.TimePeriod;

/**
 * 
 * @author Kai Reinhard (k.reinhard@micromata.de)
 * 
 */
@Entity
@Indexed
@Table(name = "T_TIMESHEET",
    indexes = {
        @javax.persistence.Index(name = "idx_fk_t_timesheet_kost2_id", columnList = "kost2_id"),
        @javax.persistence.Index(name = "idx_fk_t_timesheet_task_id", columnList = "task_id"),
        @javax.persistence.Index(name = "idx_fk_t_timesheet_user_id", columnList = "user_id"),
        @javax.persistence.Index(name = "idx_fk_t_timesheet_tenant_id", columnList = "tenant_id"),
        @javax.persistence.Index(name = "idx_timesheet_user_time", columnList = "user_id, start_time")
    })
public class TimesheetDO extends DefaultBaseDO implements Comparable<TimesheetDO>
{
  private static final long serialVersionUID = 4239370656510694224L;

  @UserPrefParameter(i18nKey = "task", orderString = "2")
  @IndexedEmbedded(depth = 1)
  private TaskDO task;

  @UserPrefParameter(i18nKey = "user", orderString = "1")
  @IndexedEmbedded(depth = 1)
  private PFUserDO user;

  private String timeZone;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.MINUTE, encoding = EncodingType.STRING)
  private Timestamp startTime;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.MINUTE, encoding = EncodingType.STRING)
  private Timestamp stopTime;

  @UserPrefParameter(i18nKey = "timesheet.location")
  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String location;

  @UserPrefParameter(i18nKey = "description", multiline = true)
  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String description;

  @UserPrefParameter(i18nKey = "fibu.kost2", orderString = "3", dependsOn = "task")
  @IndexedEmbedded(depth = 2)
  private Kost2DO kost2;

  private transient boolean marked;

  /**
   * Marker is used to mark this time sheet e. g. as a time sheet with an time period collision.
   * 
   * @return
   */
  @Transient
  public boolean isMarked()
  {
    return marked;
  }

  public TimesheetDO setMarked(final boolean marked)
  {
    this.marked = marked;
    return this;
  }

  /**
   * @return Duration in millis if startTime and stopTime is given and stopTime is after startTime, otherwise 0.
   */
  @Transient
  public long getDuration()
  {
    return getTimePeriod().getDuration();
  }

  /**
   * If this entry has a kost2 with a working time fraction set or a kost2art with a working time fraction set then the
   * fraction of millis will be returned.
   */
  @Transient
  public long getWorkFractionDuration()
  {
    if (kost2 != null) {
      if (kost2.getWorkFraction() != null) {
        return (long) (kost2.getWorkFraction().doubleValue() * getTimePeriod().getDuration());
      }
      final Kost2ArtDO kost2Art = kost2.getKost2Art();
      if (kost2Art.getWorkFraction() != null) {
        return (long) (kost2Art.getWorkFraction().doubleValue() * getTimePeriod().getDuration());
      }
    }
    return getDuration();
  }

  /**
   * @return the description
   */
  @Column(length = 4000)
  public String getDescription()
  {
    return description;
  }

  /**
   * @return The abbreviated description (maximum length is 50 characters).
   */
  @Transient
  public String getShortDescription()
  {
    if (this.description == null) {
      return "";
    }
    return StringUtils.abbreviate(getDescription(), 50);
  }

  /**
   * @param description the description to set
   */
  public TimesheetDO setDescription(final String description)
  {
    this.description = description;
    return this;
  }

  /**
   * @return the location
   */
  @Column(length = 100)
  public String getLocation()
  {
    return location;
  }

  /**
   * @param location the location to set
   */
  public TimesheetDO setLocation(final String location)
  {
    this.location = location;
    return this;
  }

  /**
   * The time zone of this time sheet in format "Europe/Berlin" etc. Not yet used!
   * 
   * @see java.util.TimeZone#getAvailableIDs()
   */
  @Column(name = "time_zone", length = 100)
  public String getTimeZone()
  {
    return timeZone;
  }

  public TimesheetDO setTimeZone(final String timeZone)
  {
    this.timeZone = timeZone;
    return this;
  }

  /**
   * @return the startTime
   */
  @Column(name = "start_time", nullable = false)
  public Timestamp getStartTime()
  {
    return startTime;
  }

  /**
   * Rounds the timestamp to DatePrecision.MINUTE_15 before.
   * 
   * @param startTime the startTime to set
   * @see DateHolder#DateHolder(java.util.Calendar, DatePrecision)
   */
  public TimesheetDO setStartTime(final Timestamp startTime)
  {
    setStartDate(startTime);
    return this;
  }

  /**
   * Rounds the timestamp to DatePrecision.MINUTE_15 before.
   * 
   * @param startTime the startTime to set
   * @see DateHolder#DateHolder(java.util.Calendar, DatePrecision)
   */
  @Transient
  public TimesheetDO setStartDate(final Date startDate)
  {
    if (startDate != null) {
      final DateHolder date = new DateHolder(startDate, DatePrecision.MINUTE_15);
      this.startTime = date.getTimestamp();
    } else {
      this.stopTime = null;
    }
    return this;
  }

  @Transient
  public TimesheetDO setStartDate(final long millis)
  {
    setStartDate(new Date(millis));
    return this;
  }

  /**
   * @return
   * @see DateTimeFormatter#formatWeekOfYear(Date)
   */
  @Transient
  public String getFormattedWeekOfYear()
  {
    return DateTimeFormatter.formatWeekOfYear(startTime);
  }

  /**
   * @return the stopTime
   */
  @Column(name = "stop_time", nullable = false)
  public Timestamp getStopTime()
  {
    return stopTime;
  }

  /**
   * @param stopTime
   * @return this for chaining.
   * @see #setStopDate(Date)
   */
  public TimesheetDO setStopTime(final Timestamp stopTime)
  {
    return setStopDate(stopTime);
  }

  /**
   * Rounds the timestamp to DatePrecision.MINUTE_15 before.
   * 
   * @param stopDate the stopTime to set
   * @return this for chaining.
   * @see DateHolder#DateHolder(java.util.Calendar, DatePrecision)
   */
  @Transient
  public TimesheetDO setStopDate(final Date stopDate)
  {
    if (stopDate != null) {
      final DateHolder date = new DateHolder(stopDate, DatePrecision.MINUTE_15);
      this.stopTime = date.getTimestamp();
    } else {
      this.stopTime = null;
    }
    return this;
  }

  @Transient
  public TimesheetDO setStopTime(final long millis)
  {
    setStopTime(new Timestamp(millis));
    return this;
  }

  @Transient
  public TimePeriod getTimePeriod()
  {
    return new TimePeriod(startTime, stopTime, marked);
  }

  /**
   * The employee assigned to this timesheet.
   * 
   * @return the user
   */
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "user_id", nullable = false)
  public PFUserDO getUser()
  {
    return user;
  }

  /**
   * @param user the user to set
   */
  public TimesheetDO setUser(final PFUserDO user)
  {
    this.user = user;
    return this;
  }

  @Transient
  public Integer getUserId()
  {
    if (this.user == null) {
      return null;
    }
    return user.getId();
  }

  /**
   * Not used as object due to performance reasons.
   * 
   * @return
   */
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "task_id", nullable = false)
  public TaskDO getTask()
  {
    return task;
  }

  public TimesheetDO setTask(final TaskDO task)
  {
    this.task = task;
    return this;
  }

  @Transient
  public Integer getTaskId()
  {
    if (this.task == null) {
      return null;
    }
    return task.getId();
  }

  @ManyToOne(fetch = FetchType.EAGER)
  @JoinColumn(name = "kost2_id", nullable = true)
  public Kost2DO getKost2()
  {
    return kost2;
  }

  public TimesheetDO setKost2(final Kost2DO kost2)
  {
    this.kost2 = kost2;
    return this;
  }

  @Transient
  public Integer getKost2Id()
  {
    if (this.kost2 == null) {
      return null;
    }
    return kost2.getId();
  }

  @Override
  public int compareTo(final TimesheetDO o)
  {
    return (getStartTime().compareTo(o.getStartTime()));
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/business/timesheet/TimesheetDO.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2025 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.timesheet

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import jakarta.persistence.*
import org.apache.commons.lang3.StringUtils
import org.hibernate.search.mapper.pojo.automaticindexing.ReindexOnUpdate
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.*
import org.projectforge.business.fibu.kost.Kost2DO
import org.projectforge.business.task.TaskDO
import org.projectforge.common.anots.PropertyInfo
import org.projectforge.common.i18n.I18nEnum
import org.projectforge.framework.calendar.DurationUtils
import org.projectforge.framework.json.IdOnlySerializer
import org.projectforge.framework.persistence.entities.DefaultBaseDO
import org.projectforge.framework.persistence.user.api.UserPrefParameter
import org.projectforge.framework.persistence.user.entities.PFUserDO
import org.projectforge.framework.time.*
import java.math.BigDecimal
import java.util.*

/**
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
@Entity
@Indexed
@Table(
    name = "T_TIMESHEET",
    indexes = [jakarta.persistence.Index(name = "idx_fk_t_timesheet_kost2_id", columnList = "kost2_id"),
        jakarta.persistence.Index(name = "idx_fk_t_timesheet_task_id", columnList = "task_id"),
        jakarta.persistence.Index(name = "idx_fk_t_timesheet_user_id", columnList = "user_id"),
        jakarta.persistence.Index(name = "idx_timesheet_user_time", columnList = "user_id, start_time")]
)
@NamedQueries(
    NamedQuery(
        name = TimesheetDO.FIND_START_STOP_BY_TASKID,
        query = "select startTime, stopTime from TimesheetDO where task.id = :taskId and deleted = false"
    ),
    NamedQuery(
        name = TimesheetDO.SELECT_MIN_MAX_DATE_FOR_USER,
        query = "select min(startTime), max(startTime) from TimesheetDO where user.id=:userId and deleted=false"
    ),
    NamedQuery(
        name = TimesheetDO.SELECT_USED_LOCATIONS_BY_USER_AND_LOCATION_SEARCHSTRING,
        query = "select distinct location from TimesheetDO where deleted=false and user.id=:userId and lastUpdate>:lastUpdate and lower(location) like :locationSearch order by location"
    ),
    NamedQuery(
        name = TimesheetDO.SELECT_RECENT_USED_LOCATIONS_BY_USER_AND_LAST_UPDATE,
        query = "select distinct location from TimesheetDO where deleted=false and user.id=:userId and lastUpdate>:lastUpdate and location!='' order by location"
    ),
    NamedQuery(
        name = TimesheetDO.SELECT_REFERENCES_BY_TASK_ID,
        query = "select distinct reference from TimesheetDO where deleted=false and task.id in :taskIds and reference is not NULL"
    )
)
open class TimesheetDO : DefaultBaseDO(), Comparable<TimesheetDO> {
    enum class TimeSavedByAIUnit(override val i18nKey: String) : I18nEnum {
        PERCENTAGE("percent"),
        HOURS("hours");
    }

    @PropertyInfo(i18nKey = "task")
    @UserPrefParameter(i18nKey = "task", orderString = "2")
    @IndexedEmbedded(includeDepth = 1)
    @IndexingDependency(reindexOnUpdate = ReindexOnUpdate.SHALLOW)
    @get:ManyToOne(fetch = FetchType.LAZY)
    @get:JoinColumn(name = "task_id", nullable = false)
    @JsonSerialize(using = IdOnlySerializer::class)
    open var task: TaskDO? = null

    @PropertyInfo(i18nKey = "user")
    @UserPrefParameter(i18nKey = "user", orderString = "1")
    @IndexedEmbedded(includeDepth = 1, includeEmbeddedObjectId = true)
    @IndexingDependency(reindexOnUpdate = ReindexOnUpdate.SHALLOW)
    @get:ManyToOne(fetch = FetchType.LAZY)
    @get:JoinColumn(name = "user_id", nullable = false)
    @JsonSerialize(using = IdOnlySerializer::class)
    open var user: PFUserDO? = null

    @get:Column(name = "time_zone", length = 100)
    open var timeZone: String? = null

    @PropertyInfo(i18nKey = "timesheet.startTime")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    //@DateBridge(resolution = Resolution.MINUTE, encoding = EncodingType.STRING)
    @get:Column(name = "start_time", nullable = false)
    open var startTime: Date? = null

    @PropertyInfo(i18nKey = "timesheet.stopTime")
    @GenericField // was:  @FullTextField(analyze = Analyze.NO)
    //@DateBridge(resolution = Resolution.MINUTE, encoding = EncodingType.STRING)
    @get:Column(name = "stop_time", nullable = false)
    open var stopTime: Date? = null

    @PropertyInfo(i18nKey = "timesheet.ai.timeSavedByAI", tooltip = "timesheet.ai.timeSavedByAI.info")
    @get:Column(name = "time_saved_by_ai")
    open var timeSavedByAI: BigDecimal? = null

    @PropertyInfo(i18nKey = "timesheet.ai.timeSavedByAIUnit", tooltip = "timesheet.ai.timeSavedByAIUnit.info")
    @get:Column(name = "time_saved_by_ai_unit")
    open var timeSavedByAIUnit: TimeSavedByAIUnit? = null

    @PropertyInfo(
        i18nKey = "timesheet.ai.timeSavedByAIDescription",
        tooltip = "timesheet.ai.timeSavedByAIDescription.info"
    )
    @get:Column(name = "time_saved_by_ai_description", length = 1000)
    open var timeSavedByAIDescription: String? = null

    @get:Transient
    val timeSavedByAIMillis: Long
        get() = AITimeSavings.getTimeSavedByAIMs(this)

    @PropertyInfo(i18nKey = "timesheet.location")
    @UserPrefParameter(i18nKey = "timesheet.location")
    @FullTextField
    @get:Column(length = 100)
    open var location: String? = null

    /**
     * Free multi purpose field.
     */
    @PropertyInfo(i18nKey = "timesheet.reference")
    @FullTextField
    @get:Column(length = 1000)
    open var reference: String? = null

    /**
     *
     */
    @PropertyInfo(i18nKey = "timesheet.tag")
    @FullTextField
    @get:Column(length = 1000)
    open var tag: String? = null

    @PropertyInfo(i18nKey = "timesheet.description")
    @UserPrefParameter(i18nKey = "description", multiline = true)
    @FullTextField
    @get:Column(length = 4000)
    open var description: String? = null

    @PropertyInfo(i18nKey = "fibu.kost2")
    @UserPrefParameter(i18nKey = "fibu.kost2", orderString = "3", dependsOn = "task")
    @IndexedEmbedded(includeDepth = 2)
    @IndexingDependency(reindexOnUpdate = ReindexOnUpdate.SHALLOW)
    @get:ManyToOne(fetch = FetchType.LAZY)
    @get:JoinColumn(name = "kost2_id", nullable = true)
    @JsonSerialize(using = IdOnlySerializer::class)
    open var kost2: Kost2DO? = null

    /**
     * Marker is used to mark this time sheet e. g. as a time sheet with an time period collision.
     */
    @get:Transient
    @JsonIgnore
    open var marked: Boolean = false

    /**
     * @return Duration in millis if startTime and stopTime is given and stopTime is after startTime, otherwise 0.
     */
    val duration: Long
        @Transient
        get() = timePeriod.duration

    /**
     * @return hours:minutes (e.g. 2:30, 12:30, ...)
     */
    val durationAsString: String
        @Transient
        get() = DurationUtils.getFormattedHoursAndMinutes(duration)

    /**
     * If this entry has a kost2 with a working time fraction set or a kost2art with a working time fraction set then the
     * fraction of millis will be returned.
     */
    val workFractionDuration: Long
        @Transient
        get() {
            if (kost2 != null) {
                if (kost2!!.workFraction != null) {
                    return (kost2!!.workFraction!!.toDouble() * timePeriod.duration).toLong()
                }
                val kost2Art = kost2!!.kost2Art
                if (kost2Art?.workFraction != null) {
                    return (kost2Art.workFraction!!.toDouble() * timePeriod.duration).toLong()
                }
            }
            return duration
        }

    /**
     * @return The abbreviated description (maximum length is 50 characters).
     */
    @Transient
    fun getShortDescription(): String? {
        return if (this.description == null)
            ""
        else
            StringUtils.abbreviate(description, 50)
    }

    /**
     * @return
     * @see DateTimeFormatter.formatWeekOfYear
     */
    @Transient
    fun getFormattedWeekOfYear(): String {
        return DateTimeFormatter.formatWeekOfYear(startTime)
    }

    val timePeriod: TimePeriod
        @Transient
        get() = TimePeriod(startTime, stopTime, marked)

    val userId: Long?
        @Transient
        get() = if (this.user == null) {
            null
        } else user!!.id

    val taskId: Long?
        @Transient
        get() = if (this.task == null) {
            null
        } else task!!.id

    val kost2Id: Long?
        @Transient
        get() = if (this.kost2 == null) {
            null
        } else kost2!!.id


    /**
     * Rounds the timestamp to DatePrecision.MINUTE_15 before.
     *
     * @param startDate the startTime to set
     * @see DateHolder#DateHolder(Date, DatePrecision)
     */
    @Transient
    fun setStartDate(startDate: Date?): TimesheetDO {
        if (startDate != null) {
            val date = PFDateTime.from(startDate).withPrecision(DatePrecision.MINUTE_5)
            this.startTime = date.sqlTimestamp
        } else {
            this.stopTime = null
        }
        return this
    }

    @Transient
    fun setStartDate(millis: Long): TimesheetDO {
        startTime = Date(millis)
        return this
    }

    /**
     * Rounds the timestamp to DatePrecision.MINUTE_15 before.
     *
     * @param stopDate the stopTime to set
     * @return this for chaining.
     * @see DateHolder#DateHolder(Date, DatePrecision)
     */
    @Transient
    fun setStopDate(stopDate: Date?): TimesheetDO {
        if (stopDate != null) {
            val date = PFDateTime.from(stopDate).withPrecision(DatePrecision.MINUTE_5)
            this.stopTime = date.sqlTimestamp
        } else {
            this.stopTime = null
        }
        return this
    }

    @Transient
    fun setStopDate(millis: Long): TimesheetDO {
        stopTime = Date(millis)
        return this
    }

    override fun compareTo(other: TimesheetDO): Int {
        return startTime?.compareTo(other.startTime) ?: 1
    }

    companion object {
        const val FIND_START_STOP_BY_TASKID = "TimesheetDO_FindStartStopByTaskId"
        internal const val SELECT_MIN_MAX_DATE_FOR_USER = "TimesheetDO_SelectMinMaxDateForUser"
        internal const val SELECT_USED_LOCATIONS_BY_USER_AND_LOCATION_SEARCHSTRING =
            "TimesheetDO_SelectLocationsByUserAndLocationSearchstring"
        internal const val SELECT_RECENT_USED_LOCATIONS_BY_USER_AND_LAST_UPDATE =
            "TimesheetDO_SelectRecentUsedLocationsByUserAndLastUpdate"
        internal const val SELECT_REFERENCES_BY_TASK_ID = "TimesheetDO_SelectReferencesByTaskId"
    }
}
